File: vane-regions/src/main/java/org/oddlama/vane/regions/RegionBlueMapLayerDelegate.java
package org.oddlama.vane.regions;

import de.bluecolored.bluemap.api.BlueMapAPI;
import de.bluecolored.bluemap.api.markers.ExtrudeMarker;
import de.bluecolored.bluemap.api.markers.MarkerSet;
import de.bluecolored.bluemap.api.math.Color;
import de.bluecolored.bluemap.api.math.Shape;
import java.util.HashMap;
import java.util.UUID;
import org.bukkit.World;
import org.bukkit.plugin.Plugin;
import org.oddlama.vane.regions.region.Region;

public class RegionBlueMapLayerDelegate {

    public static final String MARKER_SET_ID = "vane_regions.regions";

    private final RegionBlueMapLayer parent;

    private boolean bluemapEnabled = false;

    public RegionBlueMapLayerDelegate(final RegionBlueMapLayer parent) {
        this.parent = parent;
    }

    public Regions getModule() {
        return parent.getModule();
    }

    public void onEnable(final Plugin plugin) {
        BlueMapAPI.onEnable(api -> {
            getModule().log.info("Enabling BlueMap integration");
            bluemapEnabled = true;

            // Create marker sets
            for (final var world : getModule().getServer().getWorlds()) {
                createMarkerSet(api, world);
            }

            updateAllMarkers();
        });
    }

    public void onDisable() {
        if (!bluemapEnabled) {
            return;
        }

        getModule().log.info("Disabling BlueMap integration");
        bluemapEnabled = false;
    }

    // worldId -> MarkerSet
    private final HashMap<UUID, MarkerSet> markerSets = new HashMap<>();

    private void createMarkerSet(final BlueMapAPI api, final World world) {
        if (markerSets.containsKey(world.getUID())) {
            return;
        }

        final var markerSet = MarkerSet.builder()
            .label(parent.langLayerLabel.str())
            .toggleable(true)
            .defaultHidden(parent.configHideByDefault)
            .build();

        api
            .getWorld(world)
            .ifPresent(bmWorld -> {
                for (final var map : bmWorld.getMaps()) {
                    map.getMarkerSets().put(MARKER_SET_ID, markerSet);
                }
            });

        markerSets.put(world.getUID(), markerSet);
    }

    public void updateMarker(final Region region) {
        removeMarker(region.id());
        final var min = region.extent().min();
        final var max = region.extent().max();
        final var shape = Shape.createRect(min.getX(), min.getZ(), max.getX() + 1, max.getZ() + 1);

        final var marker = ExtrudeMarker.builder()
            .shape(shape, min.getY(), max.getY() + 1)
            .label(parent.langMarkerLabel.str(region.name()))
            .lineWidth(parent.configLineWidth)
            .lineColor(new Color(parent.configLineColor, (float) parent.configLineOpacity))
            .fillColor(new Color(parent.configFillColor, (float) parent.configFillOpacity))
            .depthTestEnabled(parent.configDepthTest)
            .centerPosition()
            .build();

        // Existing markers will be overwritten.
        markerSets.get(min.getWorld().getUID()).getMarkers().put(region.id().toString(), marker);
    }

    public void removeMarker(final UUID regionId) {
        for (final var markerSet : markerSets.values()) {
            markerSet.getMarkers().remove(regionId.toString());
        }
    }

    public void updateAllMarkers() {
        for (final var region : getModule().allRegions()) {
            updateMarker(region);
        }
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/RegionBlueMapLayerDelegate.kt
package org.oddlama.vane.regions

import de.bluecolored.bluemap.api.BlueMapAPI
import de.bluecolored.bluemap.api.BlueMapWorld
import de.bluecolored.bluemap.api.markers.ExtrudeMarker
import de.bluecolored.bluemap.api.markers.MarkerSet
import de.bluecolored.bluemap.api.math.Color
import de.bluecolored.bluemap.api.math.Shape
import org.bukkit.World
import org.bukkit.plugin.Plugin
import org.oddlama.vane.regions.region.Region
import java.util.*
import java.util.function.Consumer

/**
 * Low-level BlueMap API delegate used by `RegionBlueMapLayer`.
 */
class RegionBlueMapLayerDelegate(private val parent: RegionBlueMapLayer) {
    /**
     * Whether BlueMap integration is currently active.
     */
    private var bluemapEnabled = false

    /**
     * Owning regions module instance.
     */
    private val module: Regions
        get() = requireNotNull(parent.module)

    /**
     * Registers BlueMap hooks, creates marker sets, and performs initial marker sync.
     */
    fun onEnable(@Suppress("UNUSED_PARAMETER") plugin: Plugin?) {
        BlueMapAPI.onEnable { api: BlueMapAPI? ->
            val blueMapApi = api ?: return@onEnable
            module.log.info("Enabling BlueMap integration")
            bluemapEnabled = true

            // Create marker sets
            for (world in module.server.worlds) {
                createMarkerSet(blueMapApi, world)
            }
            updateAllMarkers()
        }
    }

    /**
     * Disables BlueMap integration state.
     */
    fun onDisable() {
        if (!bluemapEnabled) {
            return
        }

        module.log.info("Disabling BlueMap integration")
        bluemapEnabled = false
    }

    // worldId -> marker set
    /**
     * Marker sets keyed by world UUID.
     */
    private val markerSets = mutableMapOf<UUID, MarkerSet>()

    /**
     * Creates and registers a BlueMap marker set for a world if missing.
     */
    private fun createMarkerSet(api: BlueMapAPI, world: World) {
        if (markerSets.containsKey(world.uid)) {
            return
        }

        val markerSet = MarkerSet.builder()
            .label(parent.langLayerLabel?.str() ?: "")
            .toggleable(true)
            .defaultHidden(parent.configHideByDefault)
            .build()

        api
            .getWorld(world)
            .ifPresent(Consumer { bmWorld: BlueMapWorld? ->
                for (map in bmWorld?.maps.orEmpty()) {
                    map.markerSets[MARKER_SET_ID] = markerSet
                }
            })

        markerSets[world.uid] = markerSet
    }

    /**
     * Creates or replaces the marker for one region.
     */
    fun updateMarker(region: Region) {
        val regionId = region.id() ?: return
        val extent = region.extent() ?: return
        val min = extent.min() ?: return
        val max = extent.max() ?: return
        removeMarker(regionId)
        val shape = Shape.createRect(
            min.x.toDouble(),
            min.z.toDouble(),
            (max.x + 1).toDouble(),
            (max.z + 1).toDouble()
        )

        val marker = ExtrudeMarker.builder()
            .shape(shape, min.y.toFloat(), (max.y + 1).toFloat())
            .label(parent.langMarkerLabel?.str(region.name()) ?: region.name())
            .lineWidth(parent.configLineWidth)
            .lineColor(Color(parent.configLineColor, parent.configLineOpacity.toFloat()))
            .fillColor(Color(parent.configFillColor, parent.configFillOpacity.toFloat()))
            .depthTestEnabled(parent.configDepthTest)
            .centerPosition()
            .build()

        // Existing markers will be overwritten.
        markerSets[min.world.uid]?.markers?.set(regionId.toString(), marker)
    }

    /**
     * Removes a region marker from every world marker set.
     */
    fun removeMarker(regionId: UUID) {
        for (markerSet in markerSets.values) {
            markerSet.markers.remove(regionId.toString())
        }
    }

    /**
     * Synchronizes all BlueMap markers for loaded regions.
     */
    fun updateAllMarkers() {
        for (region in module.allRegions()) {
            val r = region ?: continue
            updateMarker(r)
        }
    }

    companion object {
        /**
         * Stable marker-set id used across all BlueMap worlds.
         */
        const val MARKER_SET_ID: String = "vane_regions.regions"
    }
}
